pre_commit/languages/golang.py

```diff
@@ -17,20 +25,92 @@
 from pre_commit.util import rmtree
 
 ENVIRONMENT_DIR = 'golangenv'
-get_default_version = helpers.basic_get_default_version
 health_check = helpers.basic_health_check
 
 
-def get_env_patch(venv: str) -> PatchesT:
+@functools.lru_cache(maxsize=1)
+def get_default_version() -> str:
+    if helpers.exe_exists('go'):
+        return 'system'
+    else:
+        return C.DEFAULT
+
+
+def get_env_patch(venv: str, language_version: str) -> PatchesT:
+    if language_version == 'system':
+        return (
+            ('PATH', (os.path.join(venv, 'bin'), os.pathsep, Var('PATH'))),
+        )
+
+    return (
+        ('GOROOT', os.path.join(venv, '.go')),
+        (
+            'PATH', (
+                os.path.join(venv, 'bin'), os.pathsep,
+                os.path.join(venv, '.go', 'bin'), os.pathsep, Var('PATH'),
+            ),
+        ),
+    )
+
+
+def _get_arch() -> str:  # pragma: no cover
+    arch = platform.machine().lower()
+    _ALIASES = {
+        'x86_64': 'amd64',
+        'i386': '386',
+        'aarch64': 'arm64',
+        'armv8': 'arm64',
+        'armv7l': 'armv6l',
+    }
+    return _ALIASES.get(arch, arch)
+
+
+@functools.lru_cache
+def _infer_go_version(version: str) -> str:
+    if version != C.DEFAULT:
+        return version
+    resp = urllib.request.urlopen('https://go.dev/dl/?mode=json')
+    return json.loads(resp.read())[0]['version'].lstrip('go')
+
+
+def _get_url(version: str) -> str:
+    os_name = platform.system().lower()
+    ext = 'zip' if os_name == 'windows' else 'tar.gz'
+    version = _infer_go_version(version)
     return (
-        ('PATH', (os.path.join(venv, 'bin'), os.pathsep, Var('PATH'))),
+        f'https://dl.google.com/go/go{version}.{os_name}-{_get_arch()}.{ext}'
     )
 
 
+def _install_go(version: str, dest: str) -> None:
+    try:
+        url = _get_url(version)
+        resp = urllib.request.urlopen(url)
+    except urllib.error.HTTPError as exec:  # pragma: no cover
+        if exec.code == 404:
+            raise ValueError(
+                'Could not find a version matching your system requirements '
+                f'(os={platform.system().lower()}; arch={_get_arch()})',
+            ) from exec
+        else:
+            raise
+    else:
+        with tempfile.NamedTemporaryFile(
+                delete=False,
+                suffix=f'_{os.path.basename(urlsplit(url).path)}',
+        ) as tmp_file:
+            shutil.copyfileobj(resp, tmp_file)
+            shutil.unpack_archive(tmp_file.name, dest)
+        shutil.move(os.path.join(dest, 'go'), os.path.join(dest, '.go'))
+
+
 @contextlib.contextmanager
-def in_env(prefix: Prefix) -> Generator[None, None, None]:
-    envdir = helpers.environment_dir(prefix, ENVIRONMENT_DIR, C.DEFAULT)
-    with envcontext(get_env_patch(envdir)):
+def in_env(
+        prefix: Prefix,
+        language_version: str,
```

Comment:
in a WIP branch I've standardized this function and with the name `version: str` (it'll also fit on one line) -- if you could do that here that'll save me a semantic conflict :D

Reply:
done!